### PR TITLE
Allow the passing of optimization options through the API

### PR DESCRIPTION
**Pull request by flomnes**
- Add an `OptimizationOptions` argument to public function `PerformSimulation`
- Add an `OptimizationOptions` argument to private methods `API::run` and `API::execute`
- In the existing tests, use the default instance for `OptimizationOptions`, that is

```cpp
struct OptimizationOptions
{
    //! Force ortools use
    bool ortoolsUsed = false;
    //! The solver name, sirius is the default
    std::string ortoolsSolver = "sirius";
    bool solverLogs = false;
    std::string solverParameters;
};
```
- Add a test with COIN
- Move headers around (remove unused logs.h causing issues, etc.)

### CODE REVIEW

File: src/tests/src/api_lib/test_api.cpp
Comment (pet-mit) [inline src/tests/src/api_lib/test_api.cpp:1]: can you add a test to test the new behaviour when we set the desired solver?

Comment (flomnes) [reply src/tests/src/api_lib/test_api.cpp:1]: See 9f5296548

File: src/solver/utils/ortools_utils.cpp [hunk 1]
```diff
@@ -354,33 +355,42 @@ std::string availableOrToolsSolversString()
     return solvers.str();
 }
 
-MPSolver* MPSolverFactory(const bool isMip, const std::string& solverName)
+static std::optional<std::string> translateSolverName(const std::string& solverName, bool isMip)
```
Comment (flomnes) [inline src/solver/utils/ortools_utils.cpp:357]: The previous code didn't handle `std::out_of_range` exceptions properly.

File: src/tests/src/api_internal/test_api.cpp [hunk 2]
```diff
@@ -96,3 +98,35 @@ BOOST_AUTO_TEST_CASE(result_contains_problems)
...
+                                                                 .solverParameters = ""};
+
+    auto results = api.run(*study_loader.get(), opt);
+
+    BOOST_CHECK(!results.antares_problems.empty());
+    BOOST_CHECK(!results.error);
+    BOOST_CHECK_EQUAL(results.antares_problems.weeklyProblems.size(), 52);
+}
+
+// Test where we use an invalid OR-Tools
```
Comment (pet-mit) [inline src/tests/src/api_internal/test_api.cpp]:
```suggestion
// Test where we use an invalid OR-Tools solver
```

File: src/tests/src/api_internal/test_api.cpp [hunk 3]
```diff
@@ -96,3 +98,35 @@ BOOST_AUTO_TEST_CASE(result_contains_problems)
...
+    Antares::API::APIInternal api;
+    auto study_loader = std::make_unique<InMemoryStudyLoader>();
+    const Antares::Solver::Optimization::OptimizationOptions opt{
+      .ortoolsUsed = true,
+      .ortoolsSolver = "this-solver-does-not-exist",
+      .solverLogs = true,
+      .solverParameters = ""};
+
+    auto shouldThrow = [&api, &study_loader, &opt] { return api.run(*study_loader.get(), opt); };
+    BOOST_CHECK_THROW(shouldThrow(), Antares::FatalError);
```
Comment (pet-mit) [inline src/tests/src/api_internal/test_api.cpp]: can you wait for #2500 to be merged and use the new unit_test_utils.h to check the message of the exception ? (see examples in test_component and test_system)

File: src/solver/utils/ortools_utils.cpp [hunk 4]
```diff
@@ -354,33 +355,42 @@ std::string availableOrToolsSolversString()
...
+                                 + " not found. Please use one of the following "
+                                 + availableOrToolsSolversString();
+    const Antares::Data::AssertionError assertionError(notFound);
 
-            throw Antares::Data::AssertionError(msg_to_throw);
-        }
+    auto internalSolverName = translateSolverName(solverName, isMip);
+    if (!internalSolverName)
+    {
+        throw assertionError;
```
Comment (pet-mit) [inline src/solver/utils/ortools_utils.cpp]: shouldn't we throw an std::invalid_argument exception here?

**Review by pet-mit — CHANGES_REQUESTED** [review]

File: src/solver/utils/ortools_utils.cpp (hunk 4 above)
Comment (a-zakir) [inline src/solver/utils/ortools_utils.cpp]: what's the point to catch the exception at l371 to finally throw another here?

Comment (flomnes) [reply src/solver/utils/ortools_utils.cpp]: That's right

Comment (flomnes) [reply src/solver/utils/ortools_utils.cpp]: To `throw` a more meaningful exception. `std::out_of_range` has no meaning from the caller's point of view.

Comment (flomnes) [reply src/solver/utils/ortools_utils.cpp]: So we now throw an `std::invalid_argument` exception !

Comment (a-zakir) [reply src/solver/utils/ortools_utils.cpp]: yes but you could do it the catch

**Review by pet-mit — APPROVED** [review]

**Pull request merged**